### ISSUE TITLE
Adding Multiple SASchedules for EV Selection

### ISSUE THREAD

**Issue by the-bay-kay** [EVerest/everest-demo#69]
## Context 

This PR builds off @shankari and I's work in Issue #64 .  This work will be done on the functional DepartureTime demos.

## High Level Goal:  

... to present the user with multiple charging profiles, which they then can select and use for their current charging session

### Subsequent Goals:
 
C++ Changes:
- [ ] Add to the Lowest Viable Power algorithm, adding alternative schedules (Maximum Viable Power, Optimized Curve, etc.)
- [ ] Generate multiple SAScheduleTuples, each from the alg. above 
- [ ] Adjusting ChargeParameterRes to return an SAScheduleList with multiple SAScheduleTuples 
 
UI Changes:
I foresee these being the more difficult changes, as we'll need to modify the Sim Loop to wait for user interaction.  I'll want to make a more thorough gameplan for this

- [ ] Pause StartEVSE process, so that we wait for the selected schedule 
- [ ] Add "pop up" to Node-RED, allowing users to select the charging profile (I believe making the UI is trivial, sending back the selected profile is likely is not -- will need to add mqtt routes)
- [ ] Resume the charging initialization with the selected schedule, continuing as normal.

**Comment by shankari** [comment]
@the-bay-kay I believe that the plan is for the "optimized curve" to be computed on the car.

1. The station will return two SASchedules in `ChargeParameterDiscoveryRes`
- `lowest_viable_power` and `maximum_supported_power`
2. The car will then look at those schedules, compute the optimized curve and send it in the `PowerDeliveryReq`

Note that the important part, per DJ and Jacob, is (2) since the car should have control of the charging schedule for the departure time. We are only implementing (1) because it lets us showcase multiple SASchedules without fully supporting tariffs (and because I want to 😄 ) So you might want to tackle (2) first, since even if we don't finish (1), we will have a clean demo.

**Comment by the-bay-kay** [comment]
> I believe that the plan is for the "optimized curve" to be computed on the car.... So you might want to tackle (2) first, since even if we don't finish (1), we will have a clean demo.

Aaah, right!  Cool cool -- sounds like a good plan!

**Comment by the-bay-kay** [comment]
Updated [Timeline](https://github.com/EVerest/everest-demo/issues/69#issue-2463912242) and Issue name to reflect the project's immediate goals.

**Comment by shankari** [comment]
note that node-red has a charts module: https://stevesnoderedguide.com/using-the-node-red-chart-node
https://flows.nodered.org/node/node-red-contrib-graphs

**Comment by the-bay-kay** [comment]
> ... including (i) plotting coefficients, and (ii) Sending an image
Let's start by investigating the first option.  As currently configured, our Node-RED *does* have a chart function, though it seems to be geared to presentation-style charts (Bar Charts, Pie Charts, etc.).  I don't believe we would be able to plot the anticipated equations.

Node-RED does have a curve plotting module we could add, that seems geared to plotting functions specifically! ([link](https://flows.nodered.org/node/node-red-contrib-curve))

> note that node-red has a charts module: ...
Likewise, this looks like a good option -- may be more flexible than the curve module, will compare the two options and go forward from there!

 
<details>
<summary> Existing Line Chart Screenshot </summary>

![image](https://github.com/user-attachments/assets/63cac830-ab93-4e6b-9e84-031e12bf271a)


</details>

**Comment by the-bay-kay** [comment]
Assessing our options:
1. The biggest downside with the `contrib-graph` module is that it does not plot to the ui (http://127.0.0.1:1880/ui):  Rather, it plots to (http://127.0.0.1:1880/dashboard/).  While this isn't a dealbreaker, setting two windows side-by-side isn't ideal (especially when we already have a modular UI).  
2. The curve-plotting algorithm unfortunately only takes a single variable as input (f(x)).  This almost certainly disqualifies it from use, unless (a) I fork the plugin and update the module, or (b) get creative with the Node-RED Flow Contexts.  

Let's take a shot at (2b) first, before exploring alternatives.  My hope is that we can insert the `dt` and `eamount` flow nodes into the function as "captured" variables, and use the function input for our "K" variable.

**Comment by shankari** [comment]
can't you also just plot a line graph and have DJ give you a set of coordinates?

**Comment by the-bay-kay** [comment]
> can't you also just plot a line graph and have DJ give you a set of coordinates?

 I was searching for an option to plot via coefficients like we discussed a few days ago, but that is an option. I'll go ahead with that approach!

**Comment by shankari** [comment]
> I was searching for an option to plot via coefficients like we discussed a few days ago, but that is an option. I'll go ahead with that approach!

Right, I think that plotting a standard quadratic curve would be more efficient, but if that is complex, we can fall back to the option that is known to work, at least for now

**Comment by the-bay-kay** [comment]
Just wasted a solid hour and a half trying to fix [this](https://discourse.nodered.org/t/dashboard-empty-in-ui/37492/13) bug... turns out the old curve module I was experimenting with bricked the UI... Anyway, we've got a basic chart running: let's write the function and generate some datapoints...
<details>
<summary> Chart Screenshot </summary>

![image](https://github.com/user-attachments/assets/cab9675c-ff0e-4193-ac2f-129fe306a1d6)
</details>

**Comment by the-bay-kay** [comment]
When attempting to plot an array of data, we do so for the points, but do not receive a line as we would expect.  My intuition says this is either (i) an issue with how we're formatting the inbound data, or (ii) an issue with how the chart is configured to accept the data.

Another minor issue with the line-chart method of plotting is that I have not figured out how to use non-timestamps within the X axis label. We can theoretically define custom labels (setting to "automatic" attempts to format as a time stamp), but it's uncertain if we the custom labels need be tied to a timestamp format like the defaults (HH:MM, for example)
<details>
<summary>Screenshots / Recordings </summary>

### Quadratic
https://github.com/user-attachments/assets/892242e3-134a-43ac-b953-978322f3dad8
### Linear
https://github.com/user-attachments/assets/dfdc8ef1-5f1b-4d2b-a5dc-6867da64f41e
### Linear w / Timestamps as X val
https://github.com/user-attachments/assets/1f64f7d1-da01-4075-8af3-3767ae444c3b
<details>

**Comment by shankari** [comment]
The x axis does represent time - the curve represents power drawn during the charge session, right?

**Comment by the-bay-kay** [comment]
>... we do not receive a line as we would expect. My intuition says this is either (i) an issue with how we're formatting the inbound data, or....

Yup, this was my problem.  Changing the way the arrays were nested, we get an OK result.  Let's add some templates for the powercurve data next
<details>
<summary> Rudimentary Graph Screenshot </summary>

![image](https://github.com/user-attachments/assets/61200b21-d821-452b-bd55-a9bd4aa7a3f1)

</details>

**Comment by the-bay-kay** [comment]
We're successfully plotting multiple curves -- let's hook up the MQTT messages now...
<details>
<summary>Curve Plotting Screenshot </summary>

![image](https://github.com/user-attachments/assets/6b3b5eb0-0e5e-4738-8a6a-860044bbfd7b)

</details>

**Comment by the-bay-kay** [comment]
Before making the changes to [everest-core/modules/EvseManager/EvseManager.cpp](https://github.com/EVerest/everest-core/blob/main/modules/EvseManager/EvseManager.cpp), let's better understand a few things, including 

1. The Algebraic Riccati Equation Linear-Quadratic Regulator ([LQR](https://underactuated.mit.edu/lqr.html)) that DJ's code will use to create the optimal curve
2. What "dials" we have available with this function
  a. (As I understand it: We have the departureTime, EAmount, and a 'scale' factor).  These will then be clamped by a min value, and PMax.)
3. What the new charging initialization sequence will look like
  a. E.g., there will be an additional step between receiving an SASchedule and initializing charging.  Likewise, we may need to adjust the schedule set by the user ("Add the clamps"), and will have to broadcast this to MQTT.  Perhaps I should sketch up a sequence diagram for this...

**Comment by the-bay-kay** [comment]
Effectively, there's two modes for our graph: "Preview", and "Actual".  Since we've got the rough preview working, let's see how we'll graph the "Actual" curve, once the optimizer is run on the SASchedule.

To re-render the graph, it seems like we receive ChargeParameterDiscoveryResponse (and likewise, the SASchedule) in [evcc/states/din_spec_states.py](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/din_spec_states.py#L343), specifically in ChargeParameterDiscovery().  My first thought is that we will want to utilize EVEREST_CTX's `publish()` method, as we are already doing with [AC_EVPowerReady](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/din_spec_states.py#L380).  That being said, the definition of this publication seems [vague](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/everest/context.py#L17)... Let's look for where this publication is received, or try to find the way we publish via MQTT.

Regardless, it's good to find this chunk of code where we find the SASchedule -- once we add DJ's optimizer, it should slot in roughly here.

While tracing these calls, I've been sketching out a rough sequence diagram to highlight the important calls within this process (attached below the cut).

<details>
<summary> In Progress Seq. Diagram </summary>

![image](https://github.com/user-attachments/assets/96332875-62c8-4747-a013-00ce5903888f)

</details>

**Comment by the-bay-kay** [comment]
An aside -- here is a sample of the powercurve graph preview, changing dynamically with the scale and departureTime -- this is with a dummy linear function for now, but switching to a different (quadratic) function should be as straight forward as swapping the function.

In testing, I into some limitations -- Node-RED freezes when attempting to chart a large number of points with a large range.  If necessary, we can limit the "sample rate" of the visualization, to reduce the number of points being charted (e.g.,  plotting the value every minute instead of every second may increase the speed of rendering more complex functions) 

<details>
<summary> Recording of PowerCurve "dummy demo" </summary>

https://github.com/user-attachments/assets/c50d1d92-b674-4da1-8e4c-1d290d5043d4


</details>

**Comment by the-bay-kay** [comment]
> ...  Let's look for where this publication is received, or try to find the way we publish via MQTT.

We know the answer to the first half, at least. We subscribe to the context publications within JsEvManager's [index.js](https://github.com/EVerest/everest-core/blob/a19bce95089dbdc75056a5497aaf7662d2655da6/modules/simulation/JsEvManager/index.js#L536).  Let's start by close reading this file to look examples of MQTT publication -- after all, JsEvManager has to receive the initial commands through a subscription at the very least.

**Comment by the-bay-kay** [comment]
While we subscribe to MQTT broadcasts in EVManager's [car_simulatorImpl.cpp](https://github.com/EVerest/everest-core/blob/a19bce95089dbdc75056a5497aaf7662d2655da6/modules/EvManager/main/car_simulatorImpl.cpp#L296), it appears we do not publish in a similar manner as EVSEManager's [evse_managerImpl.cpp](https://github.com/EVerest/everest-core/blob/a19bce95089dbdc75056a5497aaf7662d2655da6/modules/EvseManager/evse/evse_managerImpl.cpp#L345).  Since PowerCurve selection should occur within the EV ([context](https://github.com/EVerest/everest-demo/issues/69#issuecomment-2286969465)), we'll likely need to add publication to this file.  So, follow up question: how will JSEvManager communicate with car_simulatorImpl? Let's investigate...

**Comment by shankari** [comment]
>  there will be an additional step between receiving an SASchedule and initializing charging. 

This is the `PowerDeliveryReq` that the EVCC sends to the EVSE. It is already sent as part of the ISO message pair

**Comment by the-bay-kay** [comment]
Out of the [myriad of locations](https://github.com/search?q=org%3AEVerest+PowerDeliveryReq+&type=code&p=1) where we find PowerDeliveryReq, it seems that we want to add the optimizer to the async [process_message](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L1371) function within the PreCharge Class in `ext-switchev-iso15118/evcc/states/din_spec_states.py`, confirmed when comparing it to the PowerDeliveryReq() logs (screenshot below).  Let's trace from this to see where we can receive this information in Node-RED 

<details>
<summary> Logs </summary>

![image](https://github.com/user-attachments/assets/9000f059-73e3-4055-9029-5ad4760a7e14)

</details>

**Comment by shankari** [comment]
DIN != ISO, it is a precursor to ISO. I would look to see where I had to edit the departure time in my previous set of changes

**Comment by the-bay-kay** [comment]
In `din_states`, it seems we'll want to modify [process_message()](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L768), specifically around line [807](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L807).  Knowing where we want to change this, there are a few unanswered questions:

- [ ] How can we **preview** the Optimization Curve _before_ charging starts?
  - Currently, we use a set of coefficients to calculate a series of points that we then plot.  As of writing, this we are plotting a few dummy linear functions.
  - I've spent a few hours staring at DJ's code, and I don't think there is a clean way to convert this Python code to JavaScript.  We could re-implement the aspects of the `control` and `numpy` libraries that the script utilizes, but this seems like re-inventing the wheel. This is within possibility, but will bulk up the Node-RED code substantially.   (The Math.js library has some, but not all of our equivalent functions -- matrix(), etc.). 
  - I briefly explored calling a copy of the python function from Node-RED ([link](https://flows.nodered.org/flow/778859ca2503db35ff0e12341508efef/in/vL27qorhYRSn)), but I'm not sure this is an avenue we want to explore.  This would require our host machine to have python installed (*with* numpy and control), which we don't want to ask of the users.
  - Our final option is to have a sub-routine that (i) sends the coefficients to a brand new module, (ii) the module computes an example curve, and (iii) the module sends back the calculated coordinates to plot.  This seems like gilding the lily -- out of all of these options, the first seems to be the best.
- [x] Where do we add the power curve optimizer call?
  - We have the answer, as [above](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L807)
- [ ] How do we send the new curve points back to the UI for plotting?
  - Tracing where we use the context to publish [AC_EVPowerReady](https://github.com/EVerest/everest-core/blob/a19bce95089dbdc75056a5497aaf7662d2655da6/modules/simulation/JsEvManager/index.js#L505}, this publication is received in [car_simulatorImpl.cpp](https://github.com/EVerest/everest-core/blob/e3eb4d973a6077f8af7bd57a2c78d5316fb99412/modules/EvManager/main/car_simulatorImpl.cpp#L255) and our trusty [JsEvManager/index.js](https://github.com/EVerest/everest-core/blob/e3eb4d973a6077f8af7bd57a2c78d5316fb99412/modules/simulation/JsEvManager/index.js#L536).  
  - Within both of the aforementioned files, I don't see any existing MQTT publications we can use as an example. Compare this to the [evse_managerImpl.cpp](https://github.com/EVerest/everest-core/blob/e3eb4d973a6077f8af7bd57a2c78d5316fb99412/modules/EvseManager/evse/evse_managerImpl.cpp#L129), where we can see an example broadcast.  
  - I'm assuming we'll want to publish from `index.js`. The question then is how.  Do we follow the example of JsYetiSimulator's [index.js](https://github.com/EVerest/everest-core/blob/e3eb4d973a6077f8af7bd57a2c78d5316fb99412/modules/simulation/JsYetiSimulator/index.js#L923)?  Does the `mod` object perform the same?  Or, is [this](https://github.com/EVerest/everest-core/blob/e3eb4d973a6077f8af7bd57a2c78d5316fb99412/modules/simulation/JsEvManager/index.js#L246) publication in `JsEvManager` what we should use as a template? 
  - Once we know this, receiving the MQTT signal from Node-RED and plotting the points should be relatively trivial!

**Comment by the-bay-kay** [comment]
>  This seems like gilding the lily -- out of all of these options, the first seems to be the best.

After some discussion with the team, we've opted to go for "B" : Running the python script from Node-RED for the preview.  This will leave us with less overhead in the preview (we won't need to wait for the MQTT messages to bounce around), and saves the hassle of re-inventing the wheel.  

Currently, we've got basic files running on Node-RED -- an example script can be launched from the UI, and an output can be read.  The kicker is installing additional Python packages.  We need both Numpy and Control, and ideally want to upgrade to Python 3.x (Our Node-RED image is currently on 2.7).  From what I understand, we'll need to re-build the docker image, editing the dockerfile to install the correct pip packages (relevant thread [here](https://discourse.nodered.org/t/how-to-install-python-and-other-pre-requisites-for-node-red-contrib-machine-learning-v2-on-a-docker-container/47571/8)).  So, we need to (i) find the dockerfile, and (ii) re-build.  Ideally, this change should be similar to the recent changes to Everest-Manager [here](https://github.com/EVerest/everest-demo/issues/30#issuecomment-2085704770).

**Comment by the-bay-kay** [comment]
Changes have been made to `nodered/Dockerfile`, now we are (temporarily) building locally.  Now that we are doing so, we can use `apk` to install PIP, and then run pip to install the necessary dependencies.  My initial tests failed, but I believe this is because I was installing / running pip incorrectly (we install python-devel OK, but I believe need to call the correct version of pip installed with the devel package...)

<details>
<summary> File Changes </summary>

```dockerfile
# More above...
USER root
RUN apk update && \
    apk add --virtual python-dev 
USER node-red
# More below...
```


### docker-compose.ocpp201.yml
```yaml
# More above...
  nodered:
    # image: ghcr.io/everest/everest-demo/nodered:${TAG}
    build: ./nodered
# More below...
```
</details>

**Comment by shankari** [comment]
Is there no existing python install on that container? Did you try `docker exec` and then run python? Most unix distros come with python pre-installed now

**Comment by the-bay-kay** [comment]
> ... Most unix distros come with python pre-installed now

Correct!  Our docker comes pre-packaged with Python 2.7.18: the issue is, it _ does not come with pip_ (we need this to install `control` and `numpy` for the optimizer).  See:

![image](https://github.com/user-attachments/assets/335653f4-c59f-4c21-8794-74aa2d59482f)

**Comment by shankari** [comment]
why not use conda like we use elsewhere? You can download miniconda like we do in e-mission, it can set up python3 and comes with pip

**Comment by the-bay-kay** [comment]
> why not use conda like we use elsewhere? 

Not a bad idea -- that seems cleaner than doing the python package management within the Dockerfile (e.g., I've successfully updated Python, still having issues with pip...).  Let me try adding miniconda to the container now!

**Comment by the-bay-kay** [comment]
Narrowing in on the installation issue -- it seems both 2.7 and 3.9 come preinstalled (neither with pip)., the issue is that we are not correctly executing `apk add` (likely, because we are not correctly switching to root).  We should be able to install `pip3` or `conda` once we fix this.

**Comment by the-bay-kay** [comment]
When running the installation script for miniconda [link](https://docs.anaconda.com/miniconda/miniconda-install/), we run into the following issue during execution:  After accepting the license and choosing the default location, the installation terminates with the following:

![image](https://github.com/user-attachments/assets/d295a0c1-e1b0-42ac-8c8a-cff6f720a1e3)

Currently looking into what may cause this issue...

**Comment by shankari** [comment]
I tried creating the container using `docker-compose.yml` and ran into a different issue, which is related to installing on Alpine Linux https://superuser.com/questions/1621329/conda-install-in-alpine-image 

```
~ $ curl -o miniconda.sh -L https://repo.anaconda.com/miniconda/Miniconda3-py39_23.5.2-0-Linux-x86_64.sh 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 89.0M  100 89.0M    0     0  8796k      0  0:00:10  0:00:10 --:--:-- 10.2M
~ $ bash miniconda.sh -b -p $HOME
ERROR: File or directory already exists: '/usr/src/node-red'
If you want to update an existing installation, use the -u option.
~ $ bash miniconda.sh -b -p $HOME/miniconda3
PREFIX=/usr/src/node-red/miniconda3
Unpacking payload ...
miniconda.sh: line 353: /usr/src/node-red/miniconda3/conda.exe: No such file or directory
```

I can retry with the full multi-tier docker compose but let me see if I can figure it out with this first.

**Comment by shankari** [comment]
Something as simple as `python3 -m ensurepip` seems to work
Idea from: https://stackoverflow.com/a/62555259, ensurepip docs: https://docs.python.org/3/library/ensurepip.html

**Comment by the-bay-kay** [comment]
Woot!  That was the main roadblock -- we're running the script now 🥳  Let me hook up the UI inputs, and I'll push my changes and move on to the simulator portion.
<details>
<summary> Successful Script Run </summary>

![image](https://github.com/user-attachments/assets/18ed43b8-b863-48ae-b941-92090a543da1)

</details>

**Comment by the-bay-kay** [comment]
Had to futz with some of the inputs, but we now can use the UI to change these values!  There seems to be little change when the `ks` value is changed -- focusing on getting this working, we can investigate later...

Currently editing the python script to "return" a JSON object, so that the chart can read these values in the correct format!


<details>
<summary> Screen Recording of Functionality </summary>

https://github.com/user-attachments/assets/4181ef66-1b5b-4757-ab5e-b8eea3dfb833

</details>

**Comment by the-bay-kay** [comment]
We're finally plotting the curve!! The output is a bit wonky -- there are some issues plotting the `YC` component, but it _is_ being plotted.  In the interest of time, I plan to push what I have, and move on to adding the powercurve function to the simulator.  Below is a recording of the dynamic curve plotting.  Once changes are pushed to [my branch](https://github.com/the-bay-kay/everest-demo/tree/powercurve-addition), I will update with steps to reproduce.

<details>
<summary> Curve Visualization </summary>

https://github.com/user-attachments/assets/75171ed8-8d68-4e32-8592-f42c81cd9e03

</details>

**Comment by the-bay-kay** [comment]
Adding a simple `logger.info('Hello World!')` to the place where I thought we needed to add the powercurve to PowerDeliveryReq ([link](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L808)), we don't see the log within a charging session (plug-in, charging, unplug).... Perhaps I was off the mark? Let's look for where else we could add this

**Comment by the-bay-kay** [comment]
It seems the existing [build_current_demand_data](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L967) is only for DC, so if we want to modify the AC charging event, we may need to modify [this](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L949)... First, let's explore the EVCC's simulator.py, specifically [get_charge_params_v2](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/controller/simulator.py#L276C15-L276C35)

**Comment by the-bay-kay** [comment]
To get a hack-y demo working, our current progress has included:
- Generating the curve within `evcc/simulator.py`'s [process_sa_schedules_v2](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/controller/simulator.py#L575) Using the PMax and Duration_Time from `secc_schedule.p_max_schedule.schedule_entries`.
  - [ ] We will eventually need to figure out how to extract `EAmount` from these values (e.g., use PMax and DT to re-calculate the Watt-Hours  )-- at the moment, the scalar factor is hardcoded (which we have discussed), and we are hard-coding EAmount.    We'll need to shift some things around to get these values s.t. the curve can be also passed to `PowerDeliveryReq`
- We've added `paho-mqtt` back to `ext-switchev-iso15118`, in order to broadcast this updated curve to Node-RED.  (I say "back", because this repository [used to use paho-mqtt](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/CHANGELOG.md?plain=1#L457) once upon a time!) We've successfully connected this to Node-RED, all that is left is to adjust Node-RED for the visual.
  - [ ] As with the other pip additions, this will need to be added to the docker image / dockerfile

**Comment by the-bay-kay** [comment]
Some unexpected behavior: When running a `paho-mqtt` from a preview script, it behaves ok...
<details>
<summary> Proof of Concept Call </summary>

https://github.com/user-attachments/assets/69ebe7b9-c729-4837-ae78-2c76d7622e6d

</details>

But when adding the exact same call to `evcc/simulator.py`...
```python
curve_vals = LQRchargecurve(departure_time, eamount, ks)
publish.single("everest_external/nodered/{}/evcc/powercurve", "!dlrow olleH", hostname="mqtt-server")
```

We fail to start charging, with the following error (We know this is the issue, as removing the publication allows us to charge as expected -- the LQR runs OK)

<img width="1604" alt="image" src="https://github.com/user-attachments/assets/3ea42015-4591-4b9e-abcf-ca7b995024e0">

While we aren't communicating via MQTT in this module, perhaps we're crowding the outgoing messages, and have some overlap elsewhere?  Let's explore this further...

**Comment by the-bay-kay** [comment]
Finally got the curve data sent to MQTT!! Had to return the curve data with `process_sa_schedule_v2`, and then send it within `iso15118_2_states.py` (had to convert to a string) -- just need to finish the Node-RED plotting log, and then we should be good to go!

**Comment by the-bay-kay** [comment]
We've got a working demo! 🥳 

Below is a recording of the demo -- highlights include:
- Adjusting DT, EAmount, and Scale update the Curve Preview
- The PowerCurve updates when we begin charging
- While charging, we cannot change the curve -- we can only set an "ideal" curve in the "Idle" state!

There are definitely some kinks that we need to iron out, but exciting to see it working.  Once I get the relevant code committed, I'll go ahead and post steps to re-produce this demo.


https://github.com/user-attachments/assets/3f12d719-2fe6-479f-9d84-12cb9f45bc25

**Comment by the-bay-kay** [comment]
With the Node-RED visualiations working, and the powercurve correctly added, the next step is to adjust the simulation process in order to reflect the EVCC's powercurve.  

I believe the best way to do so is by generating a `ChargingProfile` for the `PowerDeliveryReq`, sent directly after the `ChargeParameterDiscoveryRes()`.  I believe this approach:

1. Follows the spirit of the ISO 15118-2, as we are not adding any responses to the PowerDeliveryReq
2. Complies with [V2G2-741] as we are allowed to change the ChargingProfile, but _not_ the SAScheduleTuple.  (Because each tuple has a unique ID that we cannot mutate after initialization, I interpret this as "we cannot mutate the tuple)
  - Likewise, [V2G2-741] Note 3 states that we may change the Charging Profile.

Below is a sequence diagram roughly outlining the flow of PowerCurve Generation & Adoption:

<details>
<summary> SequenceDiagram </summary>


<img width="927" alt="image" src="https://github.com/user-attachments/assets/82485b45-4fba-40ee-8681-c15b540ea943">


</details>

**Comment by shankari** [comment]
@the-bay-kay what is the timing of these steps? Is there actually potential to wait for the user to select a power curve?
I assume not, but I wonder if the spec only requires strict timing after the PowerDeliveryReq and whether there is room for limited user interaction that is compatible with the spec

**Comment by the-bay-kay** [comment]
> @the-bay-kay what is the timing of these steps? Is there actually potential to wait for the user to select a power curve? I assume not, but I wonder if the spec only requires strict timing after the PowerDeliveryReq and whether there is room for limited user interaction that is compatible with the spec

Good question -- as written, we immediately shoot off the PowerDeliveryReq after receiving the ChargeParameterDsicoveryRequest.  Looking at the spec...
- [V2G2-847] specifies that a state transition must occur no later than 250ms _after_ the PowerDeliveryRequest -- not an issue, since we'd set the profile before...
  - Likewise, [V2G2-866] requires an SECC to set the PWM to 100% after the PowerDeliveryReq, not relevant....

Other than these 2 clauses (additionally, V2G2-848 for the "stop" PowerDeliveryReq), it seems there are no restrictions on timings within the spec.  I see no reason why we couldn't pause for user input at this point, when we want to add algorithm selection down the road!

**Comment by the-bay-kay** [comment]
I've official got the new schedule generating, as described [above](https://github.com/EVerest/everest-demo/issues/69#issuecomment-2332603783), but am running into a timeout when attempting to send the entire schedule.  Below is a copy of the error, and the two messages being sent (the latter causing the error).  Currently investigating what may cause, this, will update accordingly...  My first thought is to try a shorter schedule, perhaps this is too much text for a single message? 

<details>
<summary> Screenshot of Error, JSON messages </summary>

### Screenshot of Timeout

![image](https://github.com/user-attachments/assets/39e5e67d-505d-43aa-b8f9-5c614c1d94e3)

### Original Message (no Timeout)
```json
{
  "V2G_Message": {
    "Header": {
      "SessionID": "11442E78D7FBFFF3"
    },
    "Body": {
      "PowerDeliveryReq": {
        "ChargeProgress": "Start",
        "SAScheduleTupleID": 1,
        "ChargingProfile": {
          "ProfileEntry": [
            {
              "ChargingProfileEntryStart": 0,
              "ChargingProfileEntryMaxPower": {
                "Value": 15000,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 8400,
              "ChargingProfileEntryMaxPower": {
                "Value": 0,
                "Multiplier": 0,
                "Unit": "W"
              }
            }
          ]
        }
      }
    }
  }
}
```

### New Schedule Msg (Timeout)
```json
{
  "V2G_Message": {
    "Header": {
      "SessionID": "81D3EF0FB732B27D"
    },
    "Body": {
      "PowerDeliveryReq": {
        "ChargeProgress": "Start",
        "SAScheduleTupleID": 1,
        "ChargingProfile": {
          "ProfileEntry": [
            {
              "ChargingProfileEntryStart": 0,
              "ChargingProfileEntryMaxPower": {
                "Value": 15000,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 60,
              "ChargingProfileEntryMaxPower": {
                "Value": 56,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 120,
              "ChargingProfileEntryMaxPower": {
                "Value": 53,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 181,
              "ChargingProfileEntryMaxPower": {
                "Value": 50,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 241,
              "ChargingProfileEntryMaxPower": {
                "Value": 47,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 302,
              "ChargingProfileEntryMaxPower": {
                "Value": 44,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 362,
              "ChargingProfileEntryMaxPower": {
                "Value": 41,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 423,
              "ChargingProfileEntryMaxPower": {
                "Value": 39,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 483,
              "ChargingProfileEntryMaxPower": {
                "Value": 36,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 543,
              "ChargingProfileEntryMaxPower": {
                "Value": 34,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 604,
              "ChargingProfileEntryMaxPower": {
                "Value": 32,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 664,
              "ChargingProfileEntryMaxPower": {
                "Value": 30,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 725,
              "ChargingProfileEntryMaxPower": {
                "Value": 29,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 785,
              "ChargingProfileEntryMaxPower": {
                "Value": 27,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 846,
              "ChargingProfileEntryMaxPower": {
                "Value": 25,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 906,
              "ChargingProfileEntryMaxPower": {
                "Value": 24,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 966,
              "ChargingProfileEntryMaxPower": {
                "Value": 22,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 1027,
              "ChargingProfileEntryMaxPower": {
                "Value": 21,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 1087,
              "ChargingProfileEntryMaxPower": {
                "Value": 20,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 1148,
              "ChargingProfileEntryMaxPower": {
                "Value": 19,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 1208,
              "ChargingProfileEntryMaxPower": {
                "Value": 17,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 1269,
              "ChargingProfileEntryMaxPower": {
                "Value": 16,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 1329,
              "ChargingProfileEntryMaxPower": {
                "Value": 15,
                "Multiplier": 0,
                "Unit": "W"
              }
            },
            {
              "ChargingProfileEntryStart": 8400,
              "ChargingProfileEntryMaxPower": {
                "Value": 0,
                "Multiplier": 0,
                "Unit": "W"
              }
            }
          ]
        }
      }
    }
  }
}
```
</details>

**Comment by the-bay-kay** [comment]
Good news -- @shankari , like we expected, we're able to interrupt the charging initialization during the `sprocess_sa_schedule_v2`.  Below is a video of this working -- just need to send the charging curve data before the interception, and the "resume" function will return the curve that we use!

https://github.com/user-attachments/assets/d0f2f817-ec36-491c-97f4-ab3e548de733

**Comment by the-bay-kay** [comment]
Looking for our options to dynamically display a choice to the user, this appears to be non-trivial.  At first I thought nodered-dashboard's [ui-notificaiton](https://dashboard.flowfuse.com/nodes/widgets/ui-notification.html) would be our answer, but this does not allow us to display selections.  I was hopeful that Node-RED's API [RED.notify](https://nodered.org/docs/api/ui/notifications/) would be the final answer, but we receive the error below when attempting to call this....  We're using Node-RED 2.2.3, so our version should be compatible with this API. Am I misunderstanding the context which it's used?  I briefly explored using an implementation similar to [this](https://discourse.nodered.org/t/looking-for-a-text-input-dialog-popup-window-similar-to-notification-widget/23838/2), but was unsuccessful (I'm assuming it was a conflict with the dashboard package we use?)  Will keep de-bugging the RED.notify method, and update with results.

I briefly 
<details>
<summary>RED.notify() TypeError</summary>

![image](https://github.com/user-attachments/assets/d4b73d75-53c1-4550-b68b-7ee344ab0e98)

</details

</details>

**Comment by the-bay-kay** [comment]
> I think the best option ...a drop down dynamically populated via the MQTT in. A notification will then prompt the user to make a selection... and we will clear the dropdown. 

Done!  I need to fix the curve preview again, but that should be straight forward.  Good news is, the mid-charge selection process is now working! Below is a video of this in action -- for the sake of our code and my sanity, I'll wrap up tomorrow after some sleep 😀 

https://github.com/user-attachments/assets/5d47c828-c8a1-4f81-8f7a-612d443c98e0

**Comment by the-bay-kay** [comment]
We've got a **functioning demo**! Below is a video of the demo being run:  the user may select from one of two "preset curves" (a stand in for our algorithm selection), and then the charge session will occur accordingly.  Users are able to preview curves during the `idle` phase, and user inputs are used when calculating the curve presets.  

While exciting, this isn't necessarily a "camera-ready" demo -- I still want to fix the notification process, for example.  After showing this off to the team tomorrow, I'll create a series of clean-up tasks for this project.

https://github.com/user-attachments/assets/2dc64598-fec1-49a3-a5a7-4fe12fb6d37a

**Comment by the-bay-kay** [comment]
## Tasks for Demo Cleanup
Now that we've got a 90% working demo, there are several steps we need to take in order to make this production worthy / "showroom" worthy.  Below is a list of the changes that need to take place in order to finish this current iteration (roughly in order):

- [ ] Fix schedule send through PowerDeliveryRequest 
  - (currently, schedule is generated w/o exchange... I seem to have broke it at some point during curve selection changes)
- [ ] Fix powercurve visualization
  - There are two ways of going about this: 
    - (i) continue to use the chart node, fixing the user scale / presentation
    - (ii) Use MatPlotLib to visualize the charts as we do in the sample script, and find a way to present images in the dashboard
- [ ] Integrate other optimizer functions beyond LQR
- [ ] Clean up code, standardize naming conventions, etc.
- [ ] Build & Push relevant docker image, reverting compose file changes
- Update the "one-line" demo script & Dockerfiles to include:
  - [ ] The updated Node-RED flows
  - [ ] Pip installation / operation for Node-RED manager

### Future Work
There are some other ways in which we can improve this functionality.  Because these projects have a much larger scale (implementing entirely new portions of ISO15118), I've separated them out into their own list.

- [ ] Adoption of multiple SASchedules
- [ ] Integration of Tariffs ("pay more to charge fast, pay less to charge slow", etc.)

**Comment by the-bay-kay** [comment]
> ... Add Option to skip mid-charge selection

Facing a seriously bizarre issue trying to implement this.  Below is a video:  TL;DR, the MQTT exchange I've set up works perfect _the first time the Node-RED UI is loaded_, and then fails subsequent charging sessions.  

I've spent a few hours trying fixes -- this issue has never occurred with the algorithm selection process, which made me think it was a timing issue.  Delaying the messages to avoid race conditions (e.g., node-red publication prior to simulator subscription), re-arranging messages, and none of it seems to work.... will update accordingly once I find the cause.

<details>
<summary>MQTT exchange issue </summary>

https://github.com/user-attachments/assets/0f5b4236-b809-4347-8a90-45be279b34a0

</details>

**Comment by the-bay-kay** [comment]
We've made the change to selecting our algorithm prior to the charging session (as opposed to during the PowerDeliveryReq handshake).  While this works _most_ of the time, we're still running into the bizarre issue described in my last comment.  

Below is a video of the behavior occurring.  First, we "stall out" on waiting for the publication of `confirmAlgorithm`.  By injecting the value -- and thus, re-publishing the MQTT message -- we successfully restart the charging process, and perform as expected.  Then, I re-delploy the Node-RED flow, and charge with the same settings - except a different algorithm.  This charging session occurs as expected.  I have ruled out the algorithm choice as being a confounding variable -- the error has occurred with both the "smooth" and "aggressive" curves.


I've included a timing chart with my current theory as to the cause of this issue.  Because of the behavior when re-publishing the message, I believe the issue may be that we subscribe to the `confirm` message after it has already been sent -- this "race condition: then causes us to stall out.  I'm not sold this is the cause, as adding a delay to the `confirmAlgorithm()` does not eliminate this bug.  Likewise, I'm confused as to why re-deploying Node-RED "fixes" the issue temporarily.  I don't have many leads yet, but I will continue to diagnose...


![image](https://github.com/user-attachments/assets/1cd8bfa4-98e3-4d6d-bd26-6bc749af316b)


https://github.com/user-attachments/assets/b97ab9c4-68cd-4b9b-8d15-d23e0e6e411b

**Comment by the-bay-kay** [comment]
I've cleaned up and uploaded the docker images to my personal fork -- once I figure out someissues with GPG I'm having, I'll push the updated scripts / demo-repo files. Once that is done, we'll close this issue and move our work to #71 !

**Comment by the-bay-kay** [comment]
@shankari , this demo should be ready for presentation.  To reproduce, you should be able to spin up my branch....

```bash
git clone https://github.com/the-bay-kay/everest-demo.git; cd everest-demo; git checkout powercurve-addition; bash demo-iso15118-2-ac-plus-ocpp.sh -r $(pwd) -b test-demo -3
```

... and start the manager as usual with `sh /ext/source/build/run-scripts/run-sil-ocpp201-pnc.sh`.*  Once the manager is successfully spun upshould be able to interact with the demo as in the videos above!  Once you've successfully reproduced this demo, feel free to close the issue. and we can continue any polishing in the next one.

*There is a slim chance the manager may need to be re-started after a fresh clone (charging will stall on the initialization sequence)...  I haven't been able to reproduce this in 3-4 attempts, but am investigating further.  Will update this comment and push the patch once I've got the fix nailed down!

**Comment by the-bay-kay** [comment]
It seems that we stall out when we process the "ChargeParameterDiscoveryRequest" (PrepareCharging)...  It appears to be an issue with how we initialize EAmount or DepartureTime within Node-RED.  That is, on a fresh clone, we stall until we set the value.  The demo works otherwise (e.g., a fresh clone where you punch in the numbers should work OK). Let me confirm this theory and I'll push an updated image!

**Comment by the-bay-kay** [comment]
@shankari I've pushed the changes that fix the PowerDeliveryRequest -- upon a fresh clone (e.g., deleting old containers & images), it is working as expected for me.  LMK if you cannot replicate this on your end!

https://github.com/user-attachments/assets/c3fa2355-1c53-4982-b84a-22d508952f9f

**Comment by shankari** [comment]
Just confirming that, after deleting all instances and all images, I am now able to see values for the curve.
I think we can close this now, and move the discussion to #71

**Comment by the-bay-kay** [comment]
> Just confirming that, after deleting all instances and all images, I am now able to see values for the curve. I think we can close this now, and move the discussion to #71

Glad it's working on your end!  I just tested Car Control + Grid Control, and everything looks A-OK there!  

I did notice what I bevel is a regression in the "no DepartureTime" behavior -- When setting the 4Hour 10A grid clamp, we're receiving a PMax of 2500W, not the expected 6900W.  I think this is because there is a default DT of 86400 somewhere -- once that's patched, I'll push the image and close this issue.
